### ISSUE TITLE
 [Bug] downloading batch w files w same name fail....

### ISSUE THREAD

**Issue by zxzzz8** [section83/MacYTDL#30]
just installed version 1.17 and tried to download a few twitch streams in a batch that all have the same name.... and it totally failed, im not sure if it was going to also try and merge them if the first was way shorter then the later ones but you can see in the attacked log:

[youtube-dl_response-Batch_download_on-2021-05-04_Tuesday_[19_124]_at_17;49;07.txt](https://github.com/section83/MacYTDL/files/6424761/youtube-dl_response-Batch_download_on-2021-05-04_Tuesday_.19_124._at_17.49.07.txt)

there are way way more files w the same name that i plan to download, so its a big issue for me, that completely disable the batch feature....
i think just enabling the user to have a name template that uses the https://github.com/ytdl-org/youtube-dl#output-template output-template options would be the best solution as suggested in https://github.com/section83/MacYTDL/issues/9 in number 7=''Enable custom file name formats to be set by the user ?'' if im not mistaking, wouldn't it?

that would also solve another very big issue i personally have, which is that i have no way of getting atm the timestamp of publication of the vod, 
and also in general i always put very complex costume file names to everything i use, so i generally find that feature really important to me in general in any software....

**Comment by section83** [comment]
I'm not entirely sure about your question so, I'll try to cover everything that seems helpful.

MaCYTDL v1.17 probably isn't very friendly with live streams.  But, it should work.  However, if you close the small downloading (or Monitor) dialog, you then need to use Activity Monitor to stop the download.  Version 1.18 will be safer – the Monitor dialog will remain visible until the live stream recording is stopped.

I've not thought it possible to download a number of streams in a batch.  I'll have to think about how to handle it properly. But, I might be confused – are you downloading live streams or the large number of finished streams ?

The youtube-dl Python script used by MacYTDL to download videos resists saving files with the same name (obviously, macOS can't store two files with the same name in the same folder).  There is a feature request to change youtube-dl so that users can choose to over-write but, the pull request has been waiting for months with no change.  youtube-dl does not merge files with the same name.  It just protects the existing file and discards what it sees as a duplicate.

In MacYTDL, normally if you try to download the same file again you will get a warning asking whether you want to over-write the existing file or save the new file with a different name.  If you choose a different name it just adds "-2" to the end of the file name.  I have assumed that when using a batch, people would not have the same file name in successive downloads.  So, I don't have the over-write logic built into the batch function.

I'll add that to my list of changes to make.  It should just save the successive files with different names.  Another idea is to enable a different download folder for each item in a batch download. I'll look at that too.

In the meantime, you can start as many downloads as you want separately, in parallel.  For each one you can specify a different download folder and so preserve the existing file names if that's what you wish.

I'll also put custom file names and publication dates onto the todo list.  They are good suggestions.  They may be in version 1.18 or v1.19 depending on how long it takes to make all the other changes I have planned.

Many thanks for this feedback.  It helps a lot to learn what features people need in MacYTDL.

Cheers.

**Comment by zxzzz8** [comment]
thanking me? well Thank You!!!! thank you very much for all this support etc ^^ .

as for my use case i hope this will explain from where im coming from and all that:

im only downloading finished twitch streams, aka ''vod''s ( =''video on demand'' if im not mistaking....).

i originally used another GUI app for yt-dl for these past-streams https://github.com/JMARyA/YoutubeDownload 
but that app no longer works for me....
i mostly used JDownloader but senes their devs never prioritys twitch vods when twitch api change i used that other gui app as backup whenever that happened until JD update, sense yt-dl seems to always fix it right away n not after few months like JD does.... 
but after the recent api change of twitch also the gui app no longer works despite me updating yt-dl, your app works tho....

but honestly if your app gets a powerfull naming template support then i think i can finally abandon JDownloader permanently as well....

as for that naming-template, i thought you can just inherit it from yt-dl coz thats what it was w the other simple looking gui app i used before.... for example one of the templates i used in that gui app was:
-=%(id)s=-=sec'Date-%(timestamp)s=-=%(uploader)s=-==%(fulltitle)s==-=%(extractor_key)s=-=%(format)s=-.%(ext)s
and it worked perfectly w twitch....

btw i hope this link show it properly but this is the list of vods im trying to download, as you can see most of them share the same name.... https://blizzz.ovh/vod/#davidangel64?2020-12-01~2021-06-01
i always download them about once a month sense twitch deletes them after 2 months, and i usually watch sometimes even a few years after they were actually got streamed, in the order of interest i want and not the order of publication, not to mention my own pace.... thats why ive been downloading these streams for years now....

**Comment by section83** [comment]
Yes, I've added custom file name template to my list for version 1.18.  It's a common feature in other GUIs and would help quite a lot.

I'm thinking of having a "Runtime" option as well as an ongoing setting.  That way, you will be able to set the template for each download separately as well as just using the saved template ongoing (which can be turned on/off).  I haven't worked out whether/how to join the current timestamp with the YTDL template parameters.  Note, that the "timestamp" template parameter in YTDL is the "UNIX timestamp of the moment the video became available" not when the download was started.

Don't know why other GUIs don't work for you.  MacYTDL really just makes it simpler to access the youtube-dl functions.  There are quite a lot of good free GUIs out there as well as commercial offerings although, I've not found another tool which can download from more sites than youtube-dl.  If downloads from Twitch fail, you can always [make a report to the YTDL developers ](https://github.com/ytdl-org/youtube-dl/issues) but make sure to follow their rules for reporting issues.

I'll look at the Pull Request shortly.  Many thanks for that – you're my first contributor !  I can make available a copy of MacYTDL without code signing.  That way, you can make your changes and test the result.  Or, you can just take the existing v1.17 and remove all the code signing.  Beware of having more than one copy of an app on your Mac at the same time. macOS can get very confused if there are two apps with the same short name and bundle identifier.

Cheers.

**Comment by zxzzz8** [comment]
that pull request was not me, but by some friend of mine irl = P
i have no idea about codes XD , im just an end-user lol = 3

**Comment by zxzzz8** [comment]
say, if the next version is gonna take a while, i hopes its not a bother, but can you make an official pre-release with just this feature? 
sense twitch keep deleting vods, and downloading them one by one and w/o an automated naming format is quite hell.... 
i mean thats why that friend made that pull-req', so ill have that feature+bug-solved quickly....

**Comment by section83** [comment]
Uh, sorry.  I'll probably release version 1.18 tomorrow (Canberra time).  I decided to publish  the changes which work and keep researching all the other bugs/features that are taking a long time to fix.  Can you provide a copy of the "batch.txt" file you are using ?  It's located in /Users/[Your home]/Library/Preferences/MacYTDL/.

Version 1.18 has option of custom file name templates which might help.  It has almost no error checking so, there is scope to do some experimenting.

**Comment by section83** [comment]
I'll upload a copy as a beta test.  I haven't finished testing but it seems to work ok.

**Comment by zxzzz8** [comment]
thank you ^^

**Comment by zxzzz8** [comment]
i can't seem to do a batch download w the naming template, as it gives me this error window: 

![Screen Shot 2021-05-16 at 01;03;00](https://user-images.githubusercontent.com/20977838/118392758-deaee500-b643-11eb-9e9f-3cb131fa50d3.png)

the naming template works for individual files tho, which still helps a lot, and i can also put them in parallel, but i still prefer the batch option that put them one after the other....


also unrelated to this error, id be really glad to be able to see my naming template in the ''settings preview'' that shows up before putting a download....

**Comment by section83** [comment]
I've uploaded a signed and notarized copy of MacYTDL with a fix for the Batch download error (a simple bug) and with proxies, cookies and template added to "Show settings before download".  The SSBDL dialog looks quite ugly but has the info.  I have a redesign of that on my list somewhere.

Let me know of any problems.

**Comment by zxzzz8** [comment]
thank you very much ^^ .

**Comment by section83** [comment]
Does the custom template feature solve your original problem with batch downloads ?  If so, I can close this issue.

**Comment by zxzzz8** [comment]
it does solve it for me personally sense im using many things in the costume template, but idk if it will work for others that don't use many differentiating variables etc, meaning, if the filename still have the same name in a batch, will it add the proper '-2' at the end of the second file and download it properly? or will it skip the second file w the same name in the batch like before....? [and also will it do it to the extra files like the thumbnail file etc as well?]

in other words, it solved my personal issue, but idk if it was solved for the original bug in general....

**Comment by section83** [comment]
Now in bug list.
